### PR TITLE
Add foreground service location support

### PR DESCRIPTION
**Pull request by Wackymax**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)
Feature: Adds support for background (Foreground service) support to Android for when the applications goes into the background while location stream is active.

This is dependent on the following PR being released:
https://github.com/Baseflow/flutter-geolocator/pull/978

### :arrow_heading_down: What is the current behavior?

Currently when the application goes into background and the OS kills the activity then location streaming stops.

### :new: What is the new behavior (if this is a feature change)?

New options have been introduced on Android that will allow the users to configure the background behaviour allow the service to continue running in the background keeping the activity alive while the location stream is active.

### :boom: Does this PR introduce a breaking change?

A new Android service is introduced which changes the behaviour of the location stream to use the new service but existing apps will not know about this change.

### :bug: Recommendations for testing

Testing the normal mode to validate the location stream is still working as expected and testing the background mode with the application running in the background.

### :memo: Links to relevant issues/docs

https://github.com/Baseflow/flutter-geolocator/issues/948
https://github.com/Baseflow/flutter-geolocator/issues/53

### :thinking: Checklist before submitting

- [X] I made sure all projects build.
- [X] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy](https://dart.dev/tools/pub/versioning).
- [X] I updated CHANGELOG.md to add a description of the change.
- [X] I followed the style guide lines ([code style guide](https://github.com/Baseflow/flutter-geolocator/blob/master/CONTRIBUTING.md)).
- [X] I updated the relevant documentation.
- [X] I rebased onto current `master`.

### CODE REVIEW

File: geolocator/pubspec.yaml [hunk 1]
```diff
@@ -26,8 +26,8 @@ dependencies:
   flutter:
     sdk: flutter
   
-  geolocator_platform_interface: ^4.0.0
-  geolocator_android: ^3.0.2
+  geolocator_platform_interface: ^4.0.0+1
```
Comment (mvanbeusekom) [inline geolocator/pubspec.yaml:29]: Version 4.0.0+1 doesn't exists, this should probably be updated to version 4.0.3.
```suggestion
  geolocator_platform_interface: ^4.0.3
```

**Review by mvanbeusekom — CHANGES_REQUESTED** [review]
To be able to publish this update we should:

[ ] Update the version number in the `pubspec.yaml`;
[ ] Add release note the the `CHANGELOG.md`;
[ ] Update the example App to include new changes;
[ ] Add documentation on how to use the new background services in the `README.md` file.

P.S. I would also like to add the iOS changes you made. They seemed really useful. If you don't have the time, do you mind if I copy it and create a PR my self?

**Comment by Wackymax** [comment]
> To be able to publish this update we should:
> 
> * [x]  Update the version number in the `pubspec.yaml`;
> * [x]  Add release note the the `CHANGELOG.md`;
> * [ ]  Update the example App to include new changes;
> * [ ]  Add documentation on how to use the new background services in the `README.md` file.
> 
> P.S. I would also like to add the iOS changes you made. They seemed really useful. If you don't have the time, do you mind if I copy it and create a PR my self?

The changes have been applied. I only updated the readme, the example app does not set any settings for Apple or Android at the moment. In terms of the iOS changes, I don't mind bringing them in. I would be happy to help out there if you feel like it will add value to the library.

**Comment by mvanbeusekom** [comment]
That would be great. I would like to merge the iOS changes first (in a separate PR) and then update this PR to include the updated geolocator_Apple package as well. 

That way we can release everything in one go.

**Comment by Wackymax** [comment]
> That would be great. I would like to merge the iOS changes first (in a separate PR) and then update this PR to include the updated geolocator_Apple package as well.
> 
> That way we can release everything in one go.

[982](https://github.com/Baseflow/flutter-geolocator/pull/982) has been created with the iOS changes.

**Comment by mvanbeusekom** [comment]
@Wackymax just released version 2.1.1 of the geolocator_apple package (thanks for that). 

Could you update this PR to reference this new version of the package and maybe add a little bit of documentation regarding the new parameter to the README.md file?

**Comment by Wackymax** [comment]
> @Wackymax just released version 2.1.1 of the geolocator_apple package (thanks for that).
> 
> Could you update this PR to reference this new version of the package and maybe add a little bit of documentation regarding the new parameter to the README.md file?

This branch has been updated with the Apple settings changes.

**Comment by Wackymax** [comment]
@mvanbeusekom it seems like something in the iOS 2.1.1 release is broken. I am not entirely sure what or how to fix it though  🤔

**Comment by mvanbeusekom** [comment]
I will have a look in an hour or so when I am behind my desk.

**Comment by Wackymax** [comment]
I have not confirmed this with the published package but it seems like the published iOS may be breaking the official geolocator package as well because we aren't pinning versions in this package. Maybe we should remove the ^ on the platform packages to prevent them from getting upgraded automatically?

**Comment by mvanbeusekom** [comment]
> I have not confirmed this with the published package but it seems like the published iOS may be breaking the official geolocator package as well because we aren't pinning versions in this package. Maybe we should remove the ^ on the platform packages to prevent them from getting upgraded automatically?

There is indeed a strange issue, I am currently investigating more detailed.

We should not remove the `^` symbol as in many cases we would like to silently release fixes to the platform specific packages without having users to update the `geolocator` plugin it self.

**Comment by mvanbeusekom** [comment]
So something strange is happening. Somehow the `.h` and `.m` files in the `ios` and `macos` folders (which are all symbolic links to the matching file in the `apple` folder lost their link attribute. 

This means that now they are just files containing a string similar to this:
```
../../../apple/Classes/Handlers/PermissionHandler.m
```

This is of course resulting in a compile error as this is not valid Objective-C code.

**Comment by mvanbeusekom** [comment]
(sorry, closed the PR by accident)

**Comment by mvanbeusekom** [comment]
I think I know what happend, I have published the `geolocator_apple` package from a Windows machine (as that is what I had close at hand yesterday). 

I didn't realise this would delete all symbolic links when publishing to pub.dev. I will publish a hotfix now using macOS machine which should fix the issue. 

Note to self: never publish Apple stuff from a Windows machine.

**Review by mvanbeusekom — APPROVED** [review]
LGTM, thanks again for all the work @Wackymax, really appreciate it.

I will merge and release the update today. Do you mind if I mention you in a tweet announcing the update (and do you have a twitter handle)?

**Comment by Wackymax** [comment]
Wow that is such a bizarre issue 😅. Glad you found it!

Thanks that is pretty awesome! It was really fun contributing, will definitely contribute again when I see use case. My twitter handle is @HennieBrink3

**Pull request merged**